CORTX-28270 Updated readme and other architectural docs. (#1416)
Seagate/cortx-test: docs/Architectural-Overview.md

```

# Hardware Configuration for test automation setup

> VMs or Physical machines with Linux flavor having 16-32 GB RAM and 8-16 logical cores would be a decent configuration for a test machine. Minimal test machine configuration is 4 GB RAM and 2 logical cores  which is suitable for development environment. The deployment view [Test Execution Deployment](Test-Execution-Deployment-View.md) discusses more about the deployment structure. The least requirement is that there should be a network connectivity of these Cloud Form Vms or physical machines to the Cortx setups in labs or public cloud services like AWS EC2 .

Reviewer: IMO, We can remove the references to CloudForm VMs etc. specific to Seagate.

Author: I have added a generic VM.

# Abbreviations and Definitions: 

* Target: A target is an application deployment under test. 

* Autobot: A program which sets up necessary environment and pre-requisites to run test automation pointing to a single target. 

* Test Execution Framework: Loosely coupled test framework module responsible for collecting test metadata from Jira and executing optimized execution ticket provided by 

Kafka. 

* Test Executor framework: Same as a test execution framework. 

> * Dist Runner: Test execution framework module loosely couple for test framework and consumer module of execution framework.  It is responsible for collecting test metadata from Jira and creating an optimized test execution plan as per Jira Test Plans. 

Reviewer 2: Minor typo correction - 
```suggestion
* Dist Runner: Test execution framework module loosely coupled for test framework and consumer module of execution framework.  It is responsible for collecting test metadata from Jira and creating an optimized test execution plan as per Jira Test Plans. 
```

* Chaos and destructive testing framework: A sub framework responsible for implementing libraries to generate destructive test scenarios and chaos in Target deployments. 

* Object Store Data Integrity framework: A sub framework which stores s3 client state and provides library to test versioned object Store data integrity interleaved across object store version upgrade and after object store downtime.  

> * HA Tests: High availability test cases which tests Consistency, Availability and Reliability of Object Store. 

Reviewer: ```suggestion
* HA Tests: High availability test cases which test Consistency, Availability and Reliability of the Object Store. 
```

* Test Client: A separate VM or Container configured with Python Environment and prerequisites Linux and Python site packages to run automated test cases.      

# Introduction: 

> Cortx-Test is distributed test execution framework comprising multiple test runners called autobots. These autobots runs heterogeneous tests of several types (functional, limits, scale, HA tests) on different targets in optimized manner and collect test results at central database (MongoDB). The test types supported are component or system level, functional tests, load tests, security tests, etc.  

Reviewer: ```suggestion
Cortx-Test is distributed test execution framework comprising of multiple test runners called autobots. These autobots run heterogeneous tests of several types (functional, limit, scale and HA tests) on different targets in optimized manner and collect test results at central database (MongoDB). The test types supported are component tests or system level, functional, load, security tests, etc.  
```

# Background: 

> For QA certification of an enterprise product, it is always a challenge to execute the test universe within a day especially since the test universe has hundreds of thousands of tests cases of several types including component, system, and performance tests. Job schedulers and CI tools like Jenkins solve this problem to an extent by providing functionality to executing multiple tests on different machines, however it is always hard to monitor numerous test executions and deploy and manage targets and complete certification of a build within a day. The algorithms and methods mentioned in the paper explain design and implementation of test executor framework which solves this problem with automatic test configuration management, parallel test execution, reporting and storage efficiency.       

Reviewer: ```suggestion
For QA certification of an enterprise product, it is always a challenge to execute the test universe within finite timeframe (typically a day) especially since the test universe has thousands of test cases of several types including component, system and performance tests. Job schedulers and CI tools like Jenkins solve this problem to an extent by providing functionality to execute multiple tests on different machines, however it is hard to monitor numerous test executions and deploy and manage targets and complete certification of a build within a day. The algorithms and methods mentioned in the paper explain design and implementation of test executor framework which solves this problem with automatic test configuration management, parallel test execution, reporting and storage efficiency.       
```

# Framework Components: 

 

The following diagram shows all components/modules present in framework.  

> *   Master autobot: This module reads test execution data from test management tool. From this test execution data, it identifies which tests need to be executed. It performs analytics on test data and divides test cases into different chucks. Those chucks are executed by individual execution autobot. 

Reviewer 2: Typo correction for 'chucks' - 
```suggestion
*   Master autobot: This module reads test execution data from test management tool. From this test execution data, it identifies which tests need to be executed. It performs analytics on test data and divides test cases into different chunks. Those chunks are executed by individual execution autobot. 
```

Author: fixed

 

*   Execution autotot: It reads the test data from Kafka channel and using that data it schedules the test execution. 

 

*   Dashboard: Dashboard is a web page where all test data is available to users. Dashboard gets all data from mongo dB using dB rest interface. DB rest interface is designed to provide access of mongo dB to different components in framework. 

 

*   Test Management Interface: This is designed to interact with any test management tool. In this framework, this framework interacts with JIRA X-ray test management tool, it does CRUD operation on test management tool. 

 

*   Logging: Logging support is provided by framework. It captures all the required logs. After test executions, framework stores all logs to nfs share, so that it can be accessible to user at any time. 

 

*   Health Check: This module checks the health of target system. If the target system is in bad condition, then there is no use of using that system for test execution. So, before starting any test, framework checks the health of the target system using this module. If the system is not healthy, then framework schedules the test execution on another healthy system. 

 

> *   Config Management: All test data which required for test execution is managed by this module. Target specific config data is stored in MongoDB, test suite specific config data is stored in Jira/database, and test specific data is stored locally. This module captures and aggregates config data and makes that data/config maps available to all tests during execution. 	 

Reviewer 2: Minor corrections - 
```suggestion
*   Config Management: All the test data which is required for test execution is managed by this module. Target specific config data is stored in MongoDB, test suite specific config data is stored in Jira/database, and test specific data is stored locally. This module captures and aggregates config data and makes that data/config maps available to all tests during execution. 	 
```

 

## Target Locking: 

Below diagram explains the flow of target locking mechanisms. This mechanism provides two types of locks: Shared and exclusive. If the request is for parallel test execution, then it will search for systems with shared lock, as existing system with parallel execution can be used to schedule additional parallel tests.  If the request is for sequential test execution, then it will search for a system with exclusive lock, which will be dedicated system for that execution. 

For any parallel test execution request, priority will be system with shared lock. If such a system is not available, then it will check availability of system with exclusive lock. This module allocates the system only if that system is healthy. 

 

 

 

 

 

## Framework Deployment Flow: 

> This framework can be deployed and used in two ways. In multi auto bot setup mode and single auto bot setup mode. 

Reviewer 2: Maybe we can remove unintentional whitespace -
```suggestion
This framework can be deployed and used in two ways. In multi autobot setup mode and single autobot setup mode. 
```

> ## Multi Auto bot's setup: 

Reviewer 2: Same as 224 - 
```suggestion
## Multi Autobot's setup: 
```

 

 

sequence 1 to 6 is explained below in points. 

Jenkins/Scheduler triggers a job which supplies parameters like targets and config to distributed executor i.e., master autobot. 

> Master Autobot runs in distributed mode and creates an execution plan which determines the sequence of test execution and which tests can be run in parallel. This execution plan entries are feed to Kafka topic cortx-test-jobs in JSON format. 

Reviewer 2: Minor typos - 
```suggestion
Master Autobot runs in distributed mode and creates an execution plan which determines the sequence of test execution and which tests can be run in parallel. This execution plan's entries are fed to Kafka topic cortx-test-jobs in JSON format. 
```

> Test runners (single autobot) are running in distributed mode on one or more virtual machines which have Kafka consumer. They belong to the same consumer group, and it guaranteed that a test will be executed only once by any of the test runners. 

Reviewer 2: Minor suggestion - 
```suggestion
Test runners (single autobot) are running in distributed mode on one or more virtual machines which have Kafka consumer. They belong to the same consumer group, and it is guaranteed that a test will be executed only once by any of the test runners. 
```

Test reporting data is saved by Restful Reporting Server to Mongo DB through Reporting client API. Reporting Client is called at the end of pytest test run with pytest reporting hook. This data is retrieved from Reporting server for further analytics and presentation. 

 

 

# Single Auto bot's setup: 

This setup is useful when there is only one target. It is designed to be used by developers for their test validation/development. This setup uses a single autobot. 

 

 

 

 

 

 

 

 

# Framework at system level: 

Below diagram shows how this framework resides in any system and shows it’s use cases to different applications. 

 

## Benefits: 

Can run on commodity H/W and light weight VMs. The run can be started on a single client VM for multiple targets. (Storage efficient) 

Framework will start parallel execution on multiple targets; depending on health of target, it will reschedule test execution for errored or failed tests. This will ensure complete test execution of identified test plan.  

Framework will have parallel execution over a single target also which will reduce the time required for test execution. (Time efficient) 

> Maximum failures will get during the first few hours of execution. 

Reviewer 2: Maybe we can reword it to "Parallel execution will reduce the time required to uncover failures during testing"
